src/main/java/org/jabref/logic/search/indexing/PostgreIndexer.java

```diff
@@ -0,0 +1,209 @@
+            preparedStatement.addBatch();
+
+            preparedStatement.executeBatch();
+        } catch (SQLException e) {
+            LOGGER.error("Could not add an entry to the index.", e);
+        }
+    }
+
+    public void removeFromIndex(Collection<BibEntry> entries, BackgroundTask<?> task) {
+        task.setTitle(Localization.lang("Removing entries from index for %0", libraryName));
+        int i = 1;
+        for (BibEntry entry : entries) {
+            if (task.isCancelled()) {
+                LOGGER.debug("Removing entries canceled");
+                return;
+            }
+            removeFromIndex(entry);
+            task.updateProgress(i, entries.size());
+            task.updateMessage(Localization.lang("%0 of %1 entries removed from the index.", i, entries.size()));
+            i++;
+        }
+    }
+
+    private void removeFromIndex(BibEntry entry) {
+        try {
+            connection.createStatement().executeUpdate("""
+                    DELETE FROM "%s"
+                    WHERE "%s" = '%s'
+                    """.formatted(tableName, PostgreConstants.ENTRY_ID, entry.getId()));
+            LOGGER.debug("Entry {} removed from index", entry.getId());
+        } catch (SQLException e) {
+            LOGGER.error("Error deleting entry from index", e);
+        }
+    }
+
+    public void updateEntry(BibEntry entry, Field field) {
+        try {
+            // Use upsert to add the field to the index if it doesn't exist, or update it if it does
+            String updateQuery = """
+            INSERT INTO "%s" ("%s", "%s", "%s")
+            VALUES (?, ?, ?)
+            ON CONFLICT ("%s", "%s") DO UPDATE
+            SET "%s" = EXCLUDED."%s"
+            """.formatted(tableName,
+                    PostgreConstants.ENTRY_ID,
+                    PostgreConstants.FIELD_NAME,
+                    PostgreConstants.FIELD_VALUE,
+                    PostgreConstants.ENTRY_ID,
+                    PostgreConstants.FIELD_NAME,
+                    PostgreConstants.FIELD_VALUE,
+                    PostgreConstants.FIELD_VALUE);
+
+            try (PreparedStatement preparedStatement = connection.prepareStatement(updateQuery)) {
+                preparedStatement.setString(1, entry.getId());
+                preparedStatement.setString(2, field.getName());
+                preparedStatement.setString(3, entry.getField(field).orElse(""));
+                preparedStatement.executeUpdate();
+                LOGGER.debug("Updated entry {} in index", entry.getId());
+            }
+        } catch (SQLException e) {
+            LOGGER.error("Error updating entry in index", e);
+        }
+    }
+
+    public void close() {
+        HeadlessExecutorService.INSTANCE.execute(this::closeIndex);
+    }
+
+    public void closeAndWait() {
+        HeadlessExecutorService.INSTANCE.executeAndWait(this::closeIndex);
+    }
+
+    private void closeIndex() {
```

Comment:
How does this method work?

Does it mean that any time I exit JabRef, all indexes are deleted. And when I open JabRef again, I reindex the library again?

Reply:
Yes. Currently for the prototype.

Maybe, it was because of unique ids etc.

Maybe, index needs only to be rebuild if .bib file changed on disk during non-usage of JabRef. (e.g., a user has used notepad++ to modify it)